# Coding Exercises (Part 1)

Now, you will have the opportunity to practice what you have learned. <br>
__Follow the instructions__ and insert your code!

The correct answer is provided below your coding cell. There you can check whether your code is correct.

If you need some help or if you want to check your code, you can also have a look at the __solutions__.

### Have Fun!

--------------------------------------------------------------------------------------------------------------

## Exercise 5: While Loops, Conditional Statements, IRR and YTM

The ABC Company has a mature project with declining revenues and increasing costs. Net Cashflows are projected as follows:

- 500 in year 1,
- 350 in year 2,
- 200 in year 3,
- 50 in year 4,
- -100 in year 5,
- -200 in year 6

1. Separate __positive__ and __negative__ cashflows. __Create two new lists__!

In [ ]:
#run the cell
cf = [500, 350, 200, 50, -100, -200]

In [ ]:
cf_pos =[]
cf_neg =[]

for i in cf: #complete the code!
  if i < 0:
    cf_neg.append(i)
  else:
    cf_pos.append(i)
print("cf_pos: {} , cf_neg: {}".format(cf_pos , cf_neg))

cf_pos: [500, 350, 200, 50] , cf_neg: [-100, -200]


2. Get the __first unprofitable year__. Your Code should return the statement __"Year ... is the first unprofitable Year!"__

In [ ]:
count_year = 0
for idx , i in enumerate(cf):
  if count_year < 1:
    if i < 0:
      count_year += 1
      print("Year {} is the first unprofitable Year!".format(idx+1))
  else:
    break


Year 5 is the first unprofitable Year!


3. Get the __last profitable year__. Your Code should return the statement __"Year ... is the last profitable Year!"__

In [ ]:
for idx,i in enumerate(cf):
  print(idx,i)

0 500
1 350
2 200
3 50
4 -100
5 -200


In [ ]:
for idx , i in enumerate(cf):
  if i > 0:
    continue
  else:
    print("Year {} is the last profitable Year!".format(idx))
    break

Year 4 is the last profitable Year!


The __modulo operator %__ returns the __remainder__ of the division between two numbers. For instance, 5 % 3 returns 2.

In [ ]:
# run the cell
5 % 3

2

In [ ]:
# run the cell
14 % 5

4

If a is divisible by b __without remainder__, a % b __returns 0__.

In [ ]:
# run the cell
a = 9
b = 3
a % b

0

4. Return the __first number n__ that is __divisible by 362 and 216__ (without remainder).<br>
Your code shall return the following statement: __"... is the first number divisible by 362 and 216!"__

In [ ]:
a = 362
b = 216

n = 1
while True: #complete the code!
  if  (n % a == 0) and (n % b == 0):
    print("{} is the first number divisible by 362 and 216!".format(n))
    break
  else:
    n += 1

39096 is the first number divisible by 362 and 216!


(result: 39096)

The ABC Company evaluates to buy an additional machine that will increase profits/cashflows by
- 50 USD in year 1,
- 75 USD in year 2 ,
- 100 USD in year 3,
- 150 USD in year 4,
- 125 USD in year 5. (each cf at period´s end)

The machine costs 450 USD (Investment in to).

5. Calculate the __Project´s IRR__ and evaluate whether ABC should pursue the project. ABCs required rate of return is 6% p.a.

In [ ]:
cf = [ 50, 75, 100, 150, 125]

In [ ]:
NPV = - 450

In [ ]:
r = 0.06 # required rate of return

In [ ]:
cf.insert(0,NPV)
cf

[-450, 50, 75, 100, 150, 125]

In [ ]:
guess_rate = 0.06
step = 0.0000001
target_npv = 0
tolerance = 0.001

while True:
  f = 1 + guess_rate
  NPV = 0
  for i in range(len(cf)):
    NPV += cf[i] / (f**i)
    diff = NPV - target_npv

  if abs(diff) > tolerance:
    if NPV < 0:
      guess_rate -= step
    else:
      guess_rate += step
  else:
    break
print(NPV, guess_rate)
print("The current YTM for XYZ´s Corporate Bond is {}%!".format(round(guess_rate,4) * 100))

-0.000875240507625108 0.031245399999173303
The current YTM for XYZ´s Corporate Bond is 3.1199999999999997%!


(result: 3.12%)

5. Would your __decision change__ if you can increse year 5 cashflows from __125 to 175__?

In [ ]:
cf[-1] = 175
cf

[-450, 50, 75, 100, 150, 175]

In [ ]:
guess_rate = 0.06
step = 0.0000001
target_npv = 0
tolerance = 0.001

while True:
  f = 1 + guess_rate
  NPV = 0
  for i in range(len(cf)):
    NPV += cf[i] / (f**i)
    diff = NPV - target_npv

  if abs(diff) > tolerance:
    if NPV < 0:
      guess_rate -= step
    else:
      guess_rate += step
  else:
    break
print(NPV, guess_rate)
print("The current YTM for XYZ´s Corporate Bond is {}%!".format(round(guess_rate,4) * 100))

-0.0009649624460621453 0.05828739999995075
The current YTM for XYZ´s Corporate Bond is 5.83%!


5. Would your __decision change__ if you can increse year 5 cashflows from __125 to 175__?
No, because the IRR is less than %6

--------------------------------------------------------------------------

6. The ABC Government issued a 30Y Government Bond one year ago with a Coupon Rate of 5.5% (annual payments in arrears). Today´s Bond Price is 97.2 (per 100 par Value). Calculate the Bond´s __current YTM__.

In [ ]:
cf = [5.5] *29
cf[-1] += cf[-1] + 100

In [ ]:
cf.insert(0,-97.2)

In [ ]:
cf

[-97.2,
 5.5,
 5.5,
 5.5,
 5.5,
 5.5,
 5.5,
 5.5,
 5.5,
 5.5,
 5.5,
 5.5,
 5.5,
 5.5,
 5.5,
 5.5,
 5.5,
 5.5,
 5.5,
 5.5,
 5.5,
 5.5,
 5.5,
 5.5,
 5.5,
 5.5,
 5.5,
 5.5,
 5.5,
 111.0]

In [ ]:
guess_rate = 0.055
step = 0.0000001
target_npv = 0
tolerance = 0.001

while True:
  f = 1 + guess_rate
  NPV = 0
  for i in range(len(cf)):
    NPV += cf[i] / (f**i)
    diff = NPV - target_npv

  if abs(diff) > tolerance:
    if NPV < 0:
      guess_rate -= step
    else:
      guess_rate += step
  else:
    break
print(NPV, guess_rate)
print("The current YTM for XYZ´s Corporate Bond is {}%!".format(round(guess_rate, 3) * 100))

0.0008978695206032228 0.05778730000008015
The current YTM for XYZ´s Corporate Bond is 5.800000000000001%!


(result: 5.70%)

# Well Done!

---------------------------------------------------------------------------------------------------------------------